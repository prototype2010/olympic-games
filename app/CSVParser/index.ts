import {readFileSync} from 'fs';
import {camelCase, zipObject} from 'lodash';
import {Runtime} from "inspector";
import ObjectPreview = module
import {type} from "os";

export class FileReader {

    static read(filePath:string) {
        try {
            return FileReader.getFileContent(filePath);
        } catch (e) {
            console.error(`Failed to parse CSV file by path ${filePath}`,e)
        }
    }

    private static getFileContent(filePath: string) {
        return readFileSync(filePath,'utf8');
    }
}

export class CSVParser {

    static parse(data : string | undefined) : Array<RawRawCSVRecord> {

        if(!data) {
            console.warn(`There is no data to parse, ${data}`);

            return;
        }

        const linesArray = CSVParser.separateByNewSring(data);
        const headers = CSVParser.extractHeaders(linesArray);
        const rows = CSVParser.extractData(linesArray);

        return CSVParser.mapHeadesToData(headers,rows);
    }

    static extractHeaders(linesArray : Array<string> = []) {
        const headers : string = linesArray[0];
        const headersArray :Array<string> = headers.split(',');

        return headersArray.map(header => camelCase(header));
    }

    static extractData(linesArray : Array<string> = []) {
        const recordsLines = linesArray.slice(1,linesArray.length -1);

        return recordsLines.map(recordsLine => recordsLine.split(','));
    }

    static mapHeadesToData(headers : Array<string>, rows : Array<Array<string>>) {

        return rows.map(row => {
            return zipObject(headers,row)
        });
    }

    static separateByNewSring(data : string) : Array<string> {
        return data.split('\r\n')
    }
}


type IndexedObject = {[index: string] : string}
type Primitive = number | string
type Nullable<T> = T | null;

enum Sex {
    M = 'M',
    F = 'F',
}

interface RawRawCSVRecord {
    id: Nullable<Primitive>,
    name: string,
    sex:  Nullable<Sex>,
    age:  string,
    height:  Nullable<Primitive>,
    weight: Nullable<Primitive>,
    team:  string,
    noc:  string,
    games:  string,
    year:  string,
    season:  string,
    city:  string,
    sport:  string,
    event:  string,
    medal:  string,
}



export class SanitizeCSV {

    static sanitizeRecordsArray(records : Array<RawRawCSVRecord>) {

        const {sanitize,isValid} = this;
        const validRecordsArray : Array<RawRawCSVRecord> = [];

        for(const record of records) {

            if(isValid(record)) {
                validRecordsArray.push(sanitize(record))
            }
        }

        return validRecordsArray;

    }

    static sanitize(record : RawRawCSVRecord) {

        const {saniziteName,saniziteParams, saniziteId} = this;

        return [saniziteParams, saniziteName, saniziteId]
            .reduce((cumulativeValue, current) => {

                return current(cumulativeValue);
            },record)
    }

    static isValid(record : RawRawCSVRecord) {
        return record && typeof record === 'object' && Object.keys(record).length > 0;
    }

    static saniziteId(record : RawRawCSVRecord) {
        const {parseInt} = this;
        const {id} = record;

        record.id = parseInt(id);

        return record;
    }

    static saniziteParams(record : RawRawCSVRecord) {

        const {parseInt} = this;
        const {height, weight} = record;


            record.height = parseInt(height);
            record.weight = parseInt(weight);

        return record;
    }

    static parseInt(value : Nullable<Primitive>) : Nullable<number> {

        const parsed = Number.parseInt(value as string);

        if(value && Number.isFinite(parsed)  && !Number.isNaN(parsed)) {
            return parsed;
        } else {
            return null;
        }
    }

    static saniziteName (record : RawRawCSVRecord) {

        return record;
    }

    static replaceAll(value : string,...args : Array<string>) {

        return args.map(stringToReplace => {

        })

    }


}

