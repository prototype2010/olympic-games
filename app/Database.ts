import sqlite3, {Database} from 'sqlite3';
import {DB_FILE_PATH} from './ConfigReader';

const SQL = sqlite3.verbose();

export class DatabaseConnection {

    private static instance : Database;

    private constructor(){}

    static getInstance() {

        if(!DatabaseConnection.instance) {
            DatabaseConnection.instance = new SQL.Database(DB_FILE_PATH,  SQL.OPEN_READWRITE,(err : Error | null) => {
                if (err) {
                    console.error(err.message);
                }
                console.log('Connected to database.....');
            });
        }

        return DatabaseConnection.instance;
    }
}