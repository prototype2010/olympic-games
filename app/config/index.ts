import {RECORD_SEP, UNIT_SEP} from "papaparse";

export const {DB_FILE_PATH,CSV_FILE_PATH} = require('dotenv').config().parsed;

export const PARSE_OPTIONS = {
    delimiter: ",",
    newline: "\r\n",
    quoteChar: '"',
    escapeChar: '"',
    header: true,
    transformHeader: undefined,
    dynamicTyping: false,
    preview: 0,
    encoding: "",
    worker: false,
    comments: false,
    step: undefined,
    error: undefined,
    download: false,
    downloadRequestHeaders: undefined,
    skipEmptyLines: false,
    chunk: undefined,
    fastMode: undefined,
    beforeFirstChunk: undefined,
    withCredentials: undefined,
    transform: undefined,
    delimitersToGuess: [',', '\t', '|', ';', RECORD_SEP, UNIT_SEP]
};
