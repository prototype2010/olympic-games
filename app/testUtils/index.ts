import { exec } from 'child_process';

export const wrapThrowable = (funcThatThrows: any, args: Array<any>) => {
  /* */
  return function() {
    funcThatThrows.apply(null, args);
  };
};

export const getSTDOUTFromChildProcess = async (resolvedFilePath: string) => {
  return new Promise<string>((res, rej) => {
    const childProcess = exec(`node --require ts-node/register ${resolvedFilePath}`, (error, stdout) => {
      if (error) {
        console.error(error.stack);
      }

      res(stdout);
    });

    childProcess.on('exit', exitCode => {
      if (exitCode !== 0) {
        rej(exitCode);
      }
    });
  });
};
