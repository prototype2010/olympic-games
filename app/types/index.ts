export enum Medal {
  'NA' = 'NA',
  Gold = 'gold',
  Silver = 'silver',
  Bronze = 'bronze',
}

export enum Sex {
  M = 'M',
  F = 'F',
}

export interface RawCSVRecord {
  id: any;
  name: any;
  sex: any;
  age: any;
  height: any;
  weight: any;
  team: any;
  noc: any;
  games: any;
  year: any;
  season: any;
  city: any;
  sport: any;
  event: any;
  medal: any;
}

export interface SanitizedOlympiadEventRecord {
  id?: number;
  name: string;
  sex?: Sex;
  age?: number;
  height?: number;
  weight?: number;
  team: string;
  noc: string;
  games: string;
  year?: number;
  season?: Season;
  city: string;
  sport: string;
  event: string;
  medal: Medal;
}

export enum Season {
  Summer = 'summer',
  Winter = 'winter',
}

export enum Table {
  GAMES = 'games',
  RESULTS = 'results',
  TEAMS = 'teams',
  SPORTS = 'sports',
  EVENTS = 'events',
  ATHLETES = 'athletes',
}

export interface Callable<T> {
  new (...args: any[]): T;
}

export interface IndexedObject {
  [index: string]: any;
}

export enum Charts {
  Medals = 'medals',
  TopTeams = 'top-teams',
}

export interface CLIExtractorDescriptor {
  priority: number;
  required: boolean;
  extractFunction: any;
  paramName: string;
}

export interface ChartConfigs {
  'top-teams': Array<CLIExtractorDescriptor>;
  medals: Array<CLIExtractorDescriptor>;
}

export interface MedalsChartParsedArgs {
  medal?: Medal;
  season: Season;
  noc: string;
}

export interface TopTeamsChartParsedArgs {
  medal?: Medal;
  season: Season;
  year?: number;
}

export interface SanitizeConfig {
  [key: string]: Array<[Function, Array<Record<string, any> | RegExp>]>;
}

export type DBSet = Array<{ [index: string]: string | number; amount: number }>;
