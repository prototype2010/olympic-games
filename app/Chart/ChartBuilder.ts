import { DBSet } from '../types';

interface ChartData {
  headers: Array<string>;
  dbSet: DBSet;
}

export class ChartBuilder {
  private static readonly MAX_BARS = 200;
  private static readonly MEDAL_SYMBOL = 'â–ˆ';

  static build(chartData: ChartData) {
    const { dbSet, headers } = chartData;

    const readyToDisplayRows = ChartBuilder.prepareRowsToDisplay(dbSet);

    ChartBuilder.displayChart([headers, ...readyToDisplayRows]);
  }

  static prepareRowsToDisplay(dbSet: DBSet) {
    const maximumMedalAmount = Math.max(...dbSet.map(({ amount }) => amount));
    const withProportionalAmount = ChartBuilder.setProportionalAmount(dbSet, maximumMedalAmount);

    return ChartBuilder.remapAmountBars(withProportionalAmount);
  }

  static setProportionalAmount(dbSet: DBSet, maximumValue: number): DBSet {
    return dbSet.map(({ amount, ...rest }) => {
      return { amount: Math.ceil((amount / maximumValue) * ChartBuilder.MAX_BARS), ...rest };
    });
  }

  static remapAmountBars(dbSet: DBSet): Array<any> {
    return dbSet.map(({ amount, ...rest }) => {
      const otherParams = Object.values({ ...rest });

      return [
        ...otherParams,
        new Array(amount)
          .fill('')
          .map(() => ChartBuilder.MEDAL_SYMBOL)
          .join(''),
      ];
    });
  }

  static displayChart(rows: Array<Array<any>>) {
    rows.forEach(row => console.log(`${'\t'}${row.join('\t')}`));

    return rows;
  }
}
