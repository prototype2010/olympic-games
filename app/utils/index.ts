import {Athlete, Event, Game, Result, Sport, Team} from "../Database/entities";
import {OlympicEvent} from "../Database/utils/OlympicEvent";
import {Callable, SanitizedCSVRecord} from "../types";


export function makeHashKey(...args : any[]) {

    const objectToBeHashed = proceedHashFuncArguments(...args);

    return Object.entries(objectToBeHashed)
        .reduce((cumulative, current) => {

            const [key,value] = current;

            if(typeof key !== 'function' && typeof key !== 'object') {
                cumulative += `${key}=${value}#`;
            }

            return cumulative;

        },'');
}

function proceedHashFuncArguments(...args : any[]) {

    if(args.length === 1 && typeof args[0] === 'object') {
        return args[0];
    } else {
        return {...args};
    }
}

function getFromHashMap<T>(map : Map<string,any>, instance : Callable) {

    const checkHasMap =  function (hashKeyArgs : Array<any>, ...callNewArgs : Array<any>) {

        const key = makeHashKey(hashKeyArgs);

        if(map.has(key)){
            return map.get(key);
        } else {
            const newObject = new instance(...callNewArgs);
            map.set(key, newObject);

            return newObject;
        }
    };

    checkHasMap.getMap = function () {
        return [...map.values()] as Array<T>;
    };

    return checkHasMap;
}

export function mapToValidDBObjects(sanitizedSCV : Array<SanitizedCSVRecord>) {

    const athletes = getFromHashMap<Athlete>(new Map(),Athlete);
    const sports = getFromHashMap<Sport>(new Map(),Sport);
    const events = getFromHashMap<Event>(new Map(),Event);
    const teams = getFromHashMap<Team>(new Map(),Team);
    const games = getFromHashMap<Game>(new Map(),Game);

    const rows : Array<OlympicEvent> =  sanitizedSCV.map(sanitizedCSVRow => {

        const {event, sport, noc, year, city, season, id} = sanitizedCSVRow;

        return new OlympicEvent(
            athletes([id],sanitizedCSVRow),
            events([event],sanitizedCSVRow),
            new Result(sanitizedCSVRow),
            sports([sport],sanitizedCSVRow),
            teams([noc],sanitizedCSVRow),
            games([year,season,city],sanitizedCSVRow)
        );
    });

    return {
        rows,
        uniqueEntries : {
            athletes : athletes.getMap(),
            sports : sports.getMap(),
            events : events.getMap(),
            teams : teams.getMap(),
            games : games.getMap(),
        }
    };
}