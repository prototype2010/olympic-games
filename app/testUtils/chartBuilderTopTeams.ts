import { ChartBuilder } from '../Chart/ChartBuilder';

ChartBuilder.build({
  headers: ['Team name', 'Medals'],
  dbSet: [
    { amount: 2, NOC: 'X' },
    { amount: 1, NOC: 'Y' },
    { amount: 0, NOC: 'Z' },
  ],
});
