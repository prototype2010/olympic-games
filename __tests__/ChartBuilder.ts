import 'jest';
import { ChartBuilder } from '../app/Chart/ChartBuilder';
import { resolve } from 'path';
import { getSTDOUTFromChildProcess } from '../app/testUtils';

describe('Chartbuilder', () => {
  describe('ChartBuilder.prepareRowsToDisplay', () => {
    test('Amount decreases properly', () => {
      const optimizedToDisplay = ChartBuilder.prepareRowsToDisplay([
        { amount: 2000, NOC: 'X' },
        { amount: 1000, NOC: 'Y' },
        { amount: 0, NOC: 'Z' },
      ]);

      expect(optimizedToDisplay).toEqual([
        { amount: 200, NOC: 'X' },
        { amount: 100, NOC: 'Y' },
        { amount: 0, NOC: 'Z' },
      ]);
    });
  });

  describe('ChartBuilder.setProportionalAmount', () => {
    test('Proportional amount decreases properly', () => {
      const decreased = ChartBuilder.setProportionalAmount(
        [
          { amount: 2000, NOC: 'X' },
          { amount: 1000, NOC: 'Y' },
          { amount: 0, NOC: 'Z' },
        ],
        2000,
      );

      expect(decreased).toEqual([
        { amount: 200, NOC: 'X' },
        { amount: 100, NOC: 'Y' },
        { amount: 0, NOC: 'Z' },
      ]);
    });

    test('Proportional amount decreases properly', () => {
      const decreased = ChartBuilder.setProportionalAmount(
        [
          { amount: 2000, NOC: 'X' },
          { amount: 1000, NOC: 'Y' },
          { amount: 0, NOC: 'Z' },
        ],
        2000,
      );

      expect(decreased).toEqual([
        { amount: 200, NOC: 'X' },
        { amount: 100, NOC: 'Y' },
        { amount: 0, NOC: 'Z' },
      ]);
    });
  });

  describe('ChartBuilder.remapAmountBars', () => {
    test('Remap to medals works well', () => {
      const remapped = ChartBuilder.remapAmountBars([
        { amount: 2, NOC: 'X' },
        { amount: 1, NOC: 'Y' },
        { amount: 0, NOC: 'Z' },
      ]);

      expect(remapped).toEqual([
        ['X', '██'],
        ['Y', '█'],
        ['Z', ''],
      ]);
    });

    describe('ChartBuilder.build', () => {
      test('Builds chart properly', async () => {
        const resolvedFilePath = resolve(__dirname, '../app/testUtils/chartBuilderTopTeams.ts');
        const result = await getSTDOUTFromChildProcess(resolvedFilePath);

        expect(result).toMatchSnapshot();
      });
    });
  });
});
