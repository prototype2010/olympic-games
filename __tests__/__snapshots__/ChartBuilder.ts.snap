// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chartbuilder ChartBuilder.remapAmountBars ChartBuilder.build Builds chart properly 1`] = `
"	Team name	Medals
	X	████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████
	Y	████████████████████████████████████████████████████████████████████████████████████████████████████
	Z	
"
`;
